### PR TITLE
6001 formant preservation

### LINKED ISSUE #6001: Formant preservation
### Your idea

Use a formant preservation technique to improve the pitch shift quality, targeting vocals. Which one and where to integrate it is open.

### Problem to be solved

Currently, pitch shifting a human voice by only a few semitones makes it sound unrealistic.
This is because the resonances of the vocal cavity are also pitched up and down, virtually reducing or enlarging the vocal cavity dimensions and making for unrealistic characters (chipmunk or Darth Vador).
Some commercial pitch shift products offer formant preservation, e.g. Ableton's clip transposition, when using the "Warp" option and the "Pro" algorithm, yields currently a convincing result.
image

### Prior art

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #6001 

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

QA:
- [x] Pitch and Speed dialog now has an "Optimize for Voice" checkbox.
- [x] Checking it noticeably (hopefully for the best when the input is voice) affects the audio result when pitch shifting up- and down.
- [x] Checking it also affects time stretching, although less noticeably, and hopefully positively when the input is voice.
- [x] Doesn't hurt the ears when used on silence, synthetic tones or other unusual signals (if you can think of any).
- [x] The checkbox state is persistent across project saving and closing.
- [x] The checkbox can be checked/unchecked during playback. (A click may be heard when doing so, but that's acceptable.)

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
You need a Windows build fix

**Comment by saintmatthieu** [comment]
> You need a Windows build fix

I have a segmentation fault on the StaffPadTimeAndPitch unit-test, which I can't reproduce locally :/

**Comment by saintmatthieu** [comment]
Found what the problem was - memory alignment. Pushed a fix commit with an explanation. Thanks unit tests 🙏

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (patrickvi2) [inline libraries/lib-time-and-pitch/FormantShifter.cpp:114]: Why?

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (patrickvi2) [inline libraries/lib-time-and-pitch/FormantShifter.cpp:99]: As you use `factor` for calculating the bounds of your iterators, it's maybe a good idea to check `factor` is in expected range.

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (patrickvi2) [inline libraries/lib-time-and-pitch/FormantShifter.cpp:144]: Shouldn't it be `1.f / env`?

Comment (patrickvi2) [inline libraries/lib-time-and-pitch/FormantShifter.cpp:144]: As `env` can still be close to infinity small, the resulting weight might get a huge number introducing noise and artifacts. I could imagine, that either limiting `1/env` or the weight would result in less audible artifacts.

**Review by patrickvi2 — APPROVED** [review]

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifter.cpp:22]: I didn't know `expected`, that's interesting, thank you. Unfortunately we still C++17, and even our CI toolchain for MacOS doesn't support the full C++17 std lib, e.g. `<filesystem>`. I don't know the exact version numbers, but the oldest MacOS version we release Audacity for cannot deal with the most recent C++ versions.
Also, we are relatively lenient in terms of safety: we document functions with `@pre` conditions, assert these preconditions, but generally don't otherwise deal with error handling. You won't find this pattern often, where the return value is checked for error. Seems to work for us 🤷 On the other hand, unit-testing is encouraged, and I have been lazy doing this time...

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifter.cpp:55]: I will assert at the beginning of `FormantShifter::Process` like I should have, given the pre-condition stated in that methods documentation. Concerning this particular method, given its narrow scope, it usually is tolerated without.

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifter.cpp:114]: A remnant of me copying chunks from the [stftPitchShift repo](https://github.com/jurihock/stftPitchShift/blob/48042d72a43287676543950691d227732a46759b/cpp/StftPitchShift/Cepster.h#L48) to get started.
It's not a great protection, because, as you commented someplace else, the weights can still end up being huge. I expect this can happen if for some reason the input signal has a frequency region of near-zero values.
I'm about to push a protection that I think makes sense: the purpose of formant preservation is to boost or attenuate spectral regions by some reasonable amount of decibels. This commit will limit the amplification to +20dB, which is plenty for our purpose and then avoid crazy amplification.

File: libraries/lib-time-and-pitch/FormantShifter.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifter.cpp:144]: Since `env` has type `float`, I expect `1` to be typecast to float, too. But sure, it doesn't hurt adding the `.f`.

Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifter.cpp:144]: Agreed, limiting the weights it will be.

**Comment by saintmatthieu** [comment]
@patrickvi2 thanks a lot for your comments ❤️ I just pushed commits to address them, feel free to re-review.

File: libraries/lib-utility/MathApprox.h
Comment (saintmatthieu) [inline libraries/lib-utility/MathApprox.h:21]: @Paul-Licameli would you be so kind to review that me pasting your Discord message here is relevant?

**Comment by saintmatthieu** [comment]
@vsverchinsky: @patrickvi2 has reviewed the DSP part of this PR. Would you be so kind to review the rest?

File: libraries/lib-utility/CMakeLists.txt
Comment (crsib) [inline libraries/lib-utility/CMakeLists.txt:35]: I am confused. Why `lib-utility` instead of `lib-math`?

Comment (saintmatthieu) [reply libraries/lib-utility/CMakeLists.txt:35]: `lib-math` depends on `lib-soxr`. I'm reluctant to let the time-and-pitch depend on this mainly because it compiles PFFFT functions which are also found in `lib-fft`, another dependency of time-and-pitch.

File: libraries/lib-time-and-pitch/FormantShifterLogVisualization.py
Comment (crsib) [inline libraries/lib-time-and-pitch/FormantShifterLogVisualization.py:11]: I think this should be moved into a separate directory, and with `requirements.txt`

File: libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h
Comment (crsib) [inline libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h:18]: Do you plan to have different implementation?

File: libraries/lib-time-and-pitch/FormantShifterLogger.h
Comment (crsib) [inline libraries/lib-time-and-pitch/FormantShifterLogger.h:39]: Why is this a part of a logger (which is a logger otherwise)?

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: I believe you only use it for tests. It's way too intrusive, TBH

File: libraries/lib-time-and-pitch/StaffPadTimeAndPitch.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPadTimeAndPitch.cpp:25]: You use `FormantShifterLogger` directly here. This happens because `GetFftSizeOverride` must not be a part of it.

I think that you should have an empty logger for production and use FormantShifterLogger only for tests. 

Configurability must not be a part of logger.

File: libraries/lib-time-and-pitch/StaffPadTimeAndPitch.h
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPadTimeAndPitch.h:27]: You do not utilize the interface again. Was there any need for it then?

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp]: We spent a lot of time and effort to optimize it, it is weird to see it go.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:47]: Paul has introduced the class for aligned vectors, hasn't he?

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp]: You go from SIMD to scalar algorithm, and I don't see any good reason for that

File: libraries/lib-utility/CMakeLists.txt
Comment (crsib) [reply libraries/lib-utility/CMakeLists.txt:35]: And why is it a problem? It doesn't introduce a circular dependency.

File: libraries/lib-utility/MathApprox.h
Comment (crsib) [inline libraries/lib-utility/MathApprox.h:32]: And probably, it should be called FastLog2 or whatever.

File: libraries/lib-utility/MapToPositiveHalfIndex.h
Comment (crsib) [inline libraries/lib-utility/MapToPositiveHalfIndex.h:24]: I don't like it, but I don't have a strong opinion, either

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:49]: I don't understand this point. Why do we need it deterministic? How much do we care about this determinism? 

Provided that `srand` modifies the global state I wonder if something like https://en.cppreference.com/w/cpp/numeric/random/linear_congruential_engine should be used instead.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:380]: I understand that this is a new code, but I will stress that it is located in a known hot point

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.h
Comment (crsib) [inline libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.h:10]: Why this include is in the header?

**Review by crsib — CHANGES_REQUESTED** [review]
I have two issues with this PR

* FormantShifterLogger is responsible for handling configuration.
  * You define it using and interface too, and I don't see why. The only place it is used - it is used directly.
* I think we should keep process_hop SIMD ready. I expect that the fix will be very simple.

There are some minor comments as well, but I will not block the PR because of them.

File: libraries/lib-utility/CMakeLists.txt
Comment (saintmatthieu) [reply libraries/lib-utility/CMakeLists.txt:35]: Not a circular dependency, but how can you be confident that it's the `lib-fft` symbols that will be used?

File: libraries/lib-time-and-pitch/FormantShifterLogVisualization.py
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogVisualization.py:11]: What is `requirements.txt` ?

File: libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h:18]: No. It ended up being a way for me to expose only relevant methods to `FormantShifter` and the rest (`NewSamplesComing` and override getters) to whoever integrates logger and shifter together (`StaffPadTimeAndPitch`).

File: libraries/lib-time-and-pitch/FormantShifterLogVisualization.py
Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogVisualization.py:11]: Your script has non-trivial dependencies. `requirements.txt` is a standard Python to define such dependencies.

File: libraries/lib-time-and-pitch/FormantShifterLogger.h
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogger.h:39]: Because it's often related by usage: these allow overriding otherwise unexposed parameters, the effect of which one might not only need to hear but also to visualize, which then really implies the logger.

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: It's convenient. What problem do you see with it?

File: libraries/lib-time-and-pitch/FormantShifterLogger.h
Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.h:39]: I rest my case; such configuration must not be a part of the logger.

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: Provided that you need a custom config in Audacity itself, why not use the "Setting"?

Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: > It's convenient. What problem do you see with it?

It is a hack that hardcodes some random path into the build for no reason.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp]: This removal is only the removal of an earlier commit in this very PR ; the older code is left untouched.
Bad things happen if I use `vo::rotate` because it expects aligned float pointers, and `pSpec + imagingBeginBin` isn't necessarily aligned.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:47]: Not necessary here, due to [this](https://github.com/audacity/audacity/pull/6068#discussion_r1521902915).

File: libraries/lib-utility/MapToPositiveHalfIndex.h
Comment (saintmatthieu) [reply libraries/lib-utility/MapToPositiveHalfIndex.h:24]: static-asserting in the header file?

Comment (crsib) [reply libraries/lib-utility/MapToPositiveHalfIndex.h:24]: Writing unit tests in a header file.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:49]: Sounds good, I'll try.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:380]: Yes, but I cannot think of a more optimized alternative, and there are extenuating circumstances:
* only gets executed for clips that are pitch-shifted downwards,
* for a number of bins that is at most a quarter of the FFT size, when down-shifted by an octave, and less for lesser shift amounts.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.h
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.h:10]: That's a remnant of me using something more sophisticated. Can be removed I think.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (crsib) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:380]: Consider me convinced:-)

**Comment by saintmatthieu** [comment]
>   * You define it using and interface too, and I don't see why. The only place it is used - it is used directly.

Not strictly speaking: `FormantShifter` only sees the abstract API.
My experimentations lead me to this solution, mixing up both logging and configuration responsibilities. I admit that it's not perfect, but I thought it could be tolerated and further clean-up work be spared. I don't really mind addressing this, if you see it as a must.

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: > Provided that you need a custom config in Audacity itself, why not use the "Setting"?

You mean as defined in `Prefs.h`? I'd still need a write mechanism.

Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: > I'd still need a write mechanism.

Can you explain please?

Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: Sure: if e.g. a method like `std::optional<int> GetFftSizeOverride()` finds `overrideFftSizeExponent.txt` in `logDir` that has a valid int value, it will use this rather than the production value. I can hear the result change on the fly by editing that file during playback. Even if I used the prefs, I'd need this functionality, that's not exposed to the user. (And in this regard it's actually not bad that `logDir` has some random string in the production build, as it makes it near impossible for the user to inadvertently have that file at that path and get a non-production parameter.)

Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: Well, you've just made it worse. During the playback, do you randomly try to open a file? I wouldn't call that a good idea at all. Open/read/close are syscalls. They are not cheap, and they can cause havoc in some scenarios. Do we want such an approach to support debug scenarios tied strictly to one machine? I say we don't.

> And in this regard it's actually not bad that logDir has some random string in the production build

Linux package maintainers will disagree. You make the build result depend on the file system structure of the computer used to build it. 

---

I'm very much confused by this approach. Probably, it doesn't cause any noticeable difference for you, but I am very much against it.

Please create a separate config for the pitch-shifter. If you want to stick to reading files for now, please make access to the files conditionally compiled. I also prefer you use an executable/data path as a root for those files.

In the future, if we want to expose these settings to users, it would be a very quick and easy operation.

File: libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/TimeAndPitch.cpp:49]: > I don't understand this point. Why do we need it deterministic? How much do we care about this determinism?

For A/B comparison before and after some algorithmic tweaks, for example. I might want to make sure that my modifications do affect the signal, so I render it once before and once after. If the process weren't deterministic I couldn't tell.

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: > You make the build result depend on the file system structure of the computer used to build it.

I don't understand how does that matters: it's intended for development and functional only for local builds. But I don't mind using an app-data directory. I was lazy but it has advantages.

> Open/read/close are syscalls. They are not cheap, and they can cause havoc in some scenarios.

I didn't have this in mind. I want to reduce conditional compilation to a maximum to avoid code becoming outdated, but if it's only just around the file opening that's completely fine.

Comment (crsib) [reply libraries/lib-time-and-pitch/FormantShifterLogger.cpp:18]: >  it's intended for development and functional only for local builds

You push it to production builds, and that matters.

> I want to reduce conditional compilation to a maximum

You push the code only needed for the feature tuning into the production build. Expensive code.

File: libraries/lib-stretching-sequence/StretchingSequence.cpp
Comment (Paul-Licameli) [inline libraries/lib-stretching-sequence/StretchingSequence.cpp:195]: What compilation warning was fixed by the first commit?

Comment (saintmatthieu) [reply libraries/lib-stretching-sequence/StretchingSequence.cpp:195]: loss of precision warning (size_t vs int or the other way round)

File: libraries/lib-time-and-pitch/CMakeLists.txt
Comment (Paul-Licameli) [inline libraries/lib-time-and-pitch/CMakeLists.txt]: This page recommends `add_compile_definitions` instead https://cmake.org/cmake/help/latest/command/add_definitions.html

You use this macro in `FormantShifterLogger.cpp`.  I would choose a different name for the macro, like THIS_DIRECTORY, just so we aren't confusing compile-time preprocessor symbols with Cmake configure-time variables.

This commit fails to compile.

```
/Users/paullicameli/Github/audacity/libraries/lib-time-and-pitch/FormantShifter.cpp:104:31: error: no member named 'getSize' in 'staffpad::audio::FourierTransform'
   const auto fftSize = mFft->getSize();
                        ~~~~~~^
```

File: libraries/lib-time-and-pitch/FormantShifterLogger.h
Comment (Paul-Licameli) [inline libraries/lib-time-and-pitch/FormantShifterLogger.h:33]: Perhaps a `ComplexFloat` alias will be convenient

File: libraries/lib-time-and-pitch/FormantShifterLogger.cpp
Comment (Paul-Licameli) [inline libraries/lib-time-and-pitch/FormantShifterLogger.cpp:74]: This unnecessary temporary vector construction makes me wish we had `std::span` instead.

Maybe instead let `PrintPythonVector` take a range of iterators as two arguments.

File: libraries/lib-time-and-pitch/TimeAndPitchExperimentalSettings.cpp
Comment (crsib) [inline libraries/lib-time-and-pitch/TimeAndPitchExperimentalSettings.cpp]: This code only works on Windows

Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/TimeAndPitchExperimentalSettings.cpp]: That's all right, I only need it there, and if someone needs it elsewhere, she can generalize it.

File: libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h
Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/FormantShifterLoggerInterface.h:18]: Update: after some clean-up there now is an alternative, dummy implementation that's injected for production.

**Review by crsib — APPROVED** [review]
I don't see any pressing issues left

**Comment by crsib** [comment]
Please note, that tests has failed, @saintmatthieu

**Comment by dozzzzer** [comment]
@saintmatthieu 

The discovered bugs are macOS only:
1. While playing, an attempt to change the clip's pitch by entering any non-zero semitones/cents value causes Audacity to freeze;
2. While playing, if semitones/cents value is other than zero and formant preservation is ON, an attempt to disable formant preservation causes a freeze as well;
3. While playing, if the Pitch and Speed dialog is opened, an attempt to close the dialog causes a freeze.

**Comment by saintmatthieu** [comment]
@dozzzzer thank you, I didn't get thread safety right and for some reason it still worked on Windows.
@crsib the last commit addresses this.

File: libraries/lib-stretching-sequence/ClipSegment.h
Comment (crsib) [inline libraries/lib-stretching-sequence/ClipSegment.h:44]: It seems that you use the mutex for making parts of the TimeAndPitchInterface to be thread safe. I don't understand, why this mutex is not in the TimeAndPitchInterface implementation in this case.

Comment (saintmatthieu) [reply libraries/lib-stretching-sequence/ClipSegment.h:44]: In `ClipSegment.cpp`, you can see that we add two callbacks for pitch shift change and formant preservation toggling. These callbacks are issued from the main thread, meanwhile the processing is on the audio thread.
Since we have the guarantee that the `StaffPadTimeAndPitch` ctor is completely executed before these callbacks are registered, it would be safe to internalize the mutex, but the safety is more straightforward to implement and for reviewers to verify this way.
The drawback with this approach is that we lock the mutex at frame rate also if there is no time stretching. I still haven't got to the bottom of whether this is a bad thing, and I was thinking maybe you could enlighten me. In our case there'd only be contention when the user changes pitch during playback.

**Review by crsib — APPROVED** [review]
I don't see blocking issues with this PR

**Pull request merged**
